crates/turbo-tasks-memory/src/edges_set.rs

```diff
@@ -0,0 +1,592 @@
+                *entry = EdgesEntry::Output;
+                true
+            }
+            EdgesEntry::ChildAndCell0(type_id) => {
+                children.push(task);
+                *entry = EdgesEntry::Cell0(*type_id);
+                true
+            }
+            EdgesEntry::ChildOutputAndCell0(type_id) => {
+                children.push(task);
+                *entry = EdgesEntry::OutputAndCell0(*type_id);
+                true
+            }
+            EdgesEntry::Complex(set) => {
+                if set.remove(&EdgeEntry::Child) {
+                    children.push(task);
+                }
+                entry.simplify();
+                !matches!(entry, EdgesEntry::Empty)
+            }
+            _ => true,
+        });
+        children
+    }
+
+    /// Removes all dependencies from the passed `dependencies` argument
+    pub(crate) fn remove_all(&mut self, dependencies: &TaskEdgesSet) {
+        self.edges.retain(|task, entry| {
+            if let Some(other) = dependencies.edges.get(task) {
+                for item in other.iter() {
+                    entry.remove(item);
+                }
+                !matches!(entry, EdgesEntry::Empty)
+            } else {
+                true
+            }
+        });
+    }
+
+    pub(crate) fn remove(&mut self, child_id: TaskEdge) -> bool {
+        let (task, edge) = child_id.task_and_edge_entry();
+        let Entry::Occupied(mut entry) = self.edges.entry(task) else {
+            return false;
+        };
+        let edge_entry = entry.get_mut();
+        if edge_entry.remove(edge) {
+            if matches!(edge_entry, EdgesEntry::Empty) {
+                entry.remove();
+            }
+            true
+        } else {
+            false
+        }
+    }
+
+    pub fn children(&self) -> impl Iterator<Item = TaskId> + '_ {
+        self.edges.iter().filter_map(|(task, entry)| match entry {
+            EdgesEntry::Child => Some(*task),
+            EdgesEntry::ChildAndOutput => Some(*task),
+            EdgesEntry::ChildAndCell0(_) => Some(*task),
+            EdgesEntry::ChildOutputAndCell0(_) => Some(*task),
+            EdgesEntry::Complex(set) => {
+                if set.contains(&EdgeEntry::Child) {
+                    Some(*task)
+                } else {
+                    None
+                }
+            }
+            _ => None,
+        })
+    }
+}
+
+impl IntoIterator for TaskEdgesSet {
+    type Item = TaskEdge;
+    type IntoIter = impl Iterator<Item = TaskEdge>;
+
+    fn into_iter(self) -> Self::IntoIter {
+        self.edges
+            .into_iter()
+            .flat_map(|(task, entry)| entry.into_iter().map(move |e| e.into_dependency(task)))
+    }
+}
+
+#[derive(Default)]
+pub struct TaskEdgesList {
+    edges: Box<[(TaskId, EdgesEntry)]>,
+}
+
+impl TaskEdgesList {
+    pub fn into_set(self) -> TaskEdgesSet {
+        TaskEdgesSet {
+            edges: Box::new(self.edges.into_vec().into_iter().collect()),
+        }
+    }
```

Comment:
How common is this operation? I bet if we were clever we could do an optimization for the case where we're converting into a `SmallVec` that avoids the iterator.